### PR TITLE
Migrate to group traits

### PR DESCRIPTION
**Pull request by str4d**
Part of #32.

### CODE REVIEW

**Comment by str4d** [comment]
Best reviewed per-commit.

**Review by ebfull — APPROVED** [review]

File: src/pasta/curves.rs [hunk 1]
```diff
@@ -796,14 +827,61 @@ macro_rules! impl_projective_curve_specific {
...
+}
+
+macro_rules! impl_projective_curve_ext {
+    ($name:ident, $name_affine:ident, $iso:ident, $base:ident, special_a0_b5) => {
+        fn hash_to_curve<'a>(domain_prefix: &'a str) -> Box<dyn Fn(&[u8]) -> Self + 'a> {
+            use super::hashtocurve;
+
+            Box::new(move |message| {
+                let mut us = [Field::zero(); 2];
+                hashtocurve::hash_to_field($name_affine::CURVE_ID, domain_prefix, message, &mut us);
```
Comment (daira) [inline src/pasta/curves.rs:842]: Does it make more sense for `CURVE_ID` to be defined on `CurveExt`? Then this macro would not require `$name_affine`.

(I think we'll also remove `BLAKE2B_PERSONALIZATION` when we switch to using SSWU for the Halo bases; I just wanted to do that separately.)

Comment (daira) [reply src/pasta/curves.rs:842]: Oh, I guess this has already merged. We can make that simplification when we switch the Halo bases.

**Comment by daira** [comment]
Post-hoc ACK.

**Pull request merged**